File: apps/go/chapter06/sort_test.go
package chapter06_test

import (
	"reflect"
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter06"
)

var sortInput = []int{6, 4, 3, 7, 1, 9, 8}
var sortExpected = []int{1, 3, 4, 6, 7, 8, 9}

func copySlice(a []int) []int {
	b := make([]int, len(a))
	copy(b, a)
	return b
}

func TestBubbleSort(t *testing.T) {
	a := copySlice(sortInput)
	chapter06.BubbleSort(a)
	if !reflect.DeepEqual(a, sortExpected) {
		t.Errorf("BubbleSort = %v; want %v", a, sortExpected)
	}
}

func TestSelectionSort(t *testing.T) {
	a := copySlice(sortInput)
	chapter06.SelectionSort(a)
	if !reflect.DeepEqual(a, sortExpected) {
		t.Errorf("SelectionSort = %v; want %v", a, sortExpected)
	}
}

func TestInsertionSort(t *testing.T) {
	a := copySlice(sortInput)
	chapter06.InsertionSort(a)
	if !reflect.DeepEqual(a, sortExpected) {
		t.Errorf("InsertionSort = %v; want %v", a, sortExpected)
	}
}

func TestQuickSort(t *testing.T) {
	a := copySlice(sortInput)
	chapter06.QuickSort(a, 0, len(a)-1)
	if !reflect.DeepEqual(a, sortExpected) {
		t.Errorf("QuickSort = %v; want %v", a, sortExpected)
	}
}

func TestMergeSort(t *testing.T) {
	a := copySlice(sortInput)
	chapter06.MergeSort(a, 0, len(a)-1)
	if !reflect.DeepEqual(a, sortExpected) {
		t.Errorf("MergeSort = %v; want %v", a, sortExpected)
	}
}

File: apps/c/chapter06/sort_test.c
#include <string.h>
#include "../include/test_helper.h"
#include "sort.h"

static void assert_sorted(int *a, int n) {
    for (int i = 0; i < n - 1; i++)
        ASSERT_TRUE(a[i] <= a[i + 1]);
}

void test_bubble_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    bubble_sort(a, 7);
    assert_sorted(a, 7);
    ASSERT_EQ_INT(1, a[0]);
}

void test_selection_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    selection_sort(a, 7);
    assert_sorted(a, 7);
}

void test_insertion_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    insertion_sort(a, 7);
    assert_sorted(a, 7);
}

void test_shell_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    shell_sort(a, 7);
    assert_sorted(a, 7);
}

void test_quick_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    quick_sort(a, 0, 6);
    assert_sorted(a, 7);
}

void test_merge_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    merge_sort(a, 7);
    assert_sorted(a, 7);
}

void test_heap_sort(void) {
    int a[] = {6, 4, 3, 7, 1, 9, 8};
    heap_sort(a, 7);
    assert_sorted(a, 7);
}

void test_counting_sort(void) {
    int a[] = {3, 1, 4, 1, 5, 9, 2, 6};
    counting_sort(a, 8, 10);
    assert_sorted(a, 8);
    ASSERT_EQ_INT(1, a[0]);
}

int main(void) {
    test_bubble_sort();
    test_selection_sort();
    test_insertion_sort();
    test_shell_sort();
    test_quick_sort();
    test_merge_sort();
    test_heap_sort();
    test_counting_sort();
    TEST_SUMMARY();
    return 0;
}
